kvstore/src/sstable.rs

```diff
@@ -308,31 +286,41 @@ where
 
     fn next(&mut self) -> Option<Self::Item> {
         while !self.heads.is_empty() {
+            // get new key
             let (key, source_idx) = *self.heads.keys().next().unwrap();
-            let val = self.heads.remove(&(key, source_idx)).unwrap();
+            let mut val = self.heads.remove(&(key, source_idx)).unwrap();
 
             // replace
             if let Some((k, v)) = self.sources[source_idx].next() {
                 self.heads.insert((k, source_idx), v);
             }
 
-            // merge logic
-            // if deleted, remember
-            let (deleted, stale) = match self.seen.get(&key) {
-                Some(&seen_ts) if seen_ts < val.ts => {
-                    // fresh val
-                    self.seen.insert(key, val.ts);
-                    (val.val.is_none(), false)
+            // check for other versions of this record
+            while !self.heads.is_empty() {
+                let (next_key, source_idx) = *self.heads.keys().next().unwrap();
+
+                // Found a different version of the record
+                if key == next_key {
+                    // pop this version, check if it's newer
+                    let other_version = self.heads.remove(&(next_key, source_idx)).unwrap();
+                    if other_version.ts > val.ts {
+                        val = other_version;
+                    }
+
+                    // replace
+                    if let Some((k, v)) = self.sources[source_idx].next() {
+                        self.heads.insert((k, source_idx), v);
+                    }
+                } else {
+                    break;
                 }
-                Some(_) => (val.val.is_none(), true),
-                None => {
-                    self.seen.insert(key, val.ts);
-                    (val.val.is_none(), false)
-                }
-            };
+            }
 
-            if !(stale || deleted) {
+            // Don't produce deleted records
+            if val.val.is_some() {
                 return Some((key, val));
+            } else {
+                continue;
```

Comment:
looks like we're already at the bottom of the loop here so why the explicit `continue`?

Reply:
Removing it